File: iosApp/iosApp/Component/PokemonCard.swift

//
// Created by Alif Phincon on 26/09/25.
//

import ComposeApp
import SwiftUI

struct PokemonCard: View {
    let name: String
    let imageURL: String?
    let idTag: String
    
    @State private var uiImage: UIImage? = nil
    
    private let loader = SharedImageLoader(httpClient: KtorClientFactory().create())
    
    var body: some View {
        ZStack(alignment: .center) {
            VStack(spacing: 0) {
                Spacer()
                ZStack(alignment: .bottom) {
                    Rectangle()
                        .fill(Color(AppColor.background))
                        .frame(height: 44, alignment: .bottom)
//                        .cornerRadius(7, corners: [.topLeft, .topRight])
                    
                    Text(name)
                        .frame(alignment: .center)
                        .font(Font(kmmFont: FontUtils.shared.body3))
                        .foregroundColor(Color(AppColor.grayscaledark))
                        .padding(EdgeInsets(top: 0, leading: 8, bottom: 4, trailing: 8))
                }
            }
            
            if let img = uiImage {
                Image(uiImage: img)
                    .resizable()
                    .scaledToFit()
                    .frame(width: 60, height: 60)
            } else {
                ProgressView()
                    .frame(width: 60, height: 60)
                    .onAppear {
                        loadImage()
                    }
            }
            
            VStack {
                HStack {
                    Spacer()
                    HStack(spacing: 0) {
                        Image("tag")
                            .resizable()
                            .renderingMode(.template)
                            .frame(width: 12, height: 12)
                            .foregroundColor(Color(AppColor.grayscalemedium))
                        Text(String(format: "%03d", Int(idTag) ?? 0))
                            .font(Font(kmmFont: FontUtils.shared.caption))
                            .frame(height: 12)
                            .foregroundColor(Color(AppColor.grayscalemedium))
                    }
                    .padding(EdgeInsets(top: 6, leading: 0, bottom: 0, trailing: 10))
                }
                Spacer()
            }
        }
        .frame(height: 108)
        .padding(0)
        .background(
            RoundedRectangle(cornerRadius: 8)
                .fill(Color.white)
        )
        .clipShape(
            RoundedRectangle(cornerRadius: 8, style: .continuous)
        )
        .shadow(color: Color(AppColor.grayscalemedium),
                radius: 2, x: 0, y: 1)
    }
    
    private func loadImage() {
        guard let url = imageURL else { return }
        loader.load(url: url) { data, error in
            if let byteArray = data {
                let nsData = byteArray.toData()
                if let img = UIImage(data: nsData) {
                    DispatchQueue.main.async {
                        self.uiImage = img
                    }
                }
            }
        }
    }
}

File: composeApp/src/androidMain/kotlin/com/appwork/pokeapi_kmm/compose/PokemonCard.kt
package com.appwork.pokeapi_kmm.compose

import android.graphics.BitmapFactory
import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.shadow
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.appwork.pokeapi_kmm.R
import com.appwork.pokeapi_kmm.data.SharedImageLoader
import com.appwork.pokeapi_kmm.services.KtorClientFactory
import com.appwork.pokeapi_kmm.utils.AppColor
import com.appwork.pokeapi_kmm.utils.FontUtils
import com.appwork.pokeapi_kmm.utils.toComposeColor
import kotlinx.coroutines.launch
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.tooling.preview.Preview


@Composable
fun PokemonCard(
    name: String,
    imageURL: String?,
    idTag: String,
    modifier: Modifier = Modifier
) {
    // State for image bitmap
    var bitmap by remember { mutableStateOf<android.graphics.Bitmap?>(null) }
    val scope = rememberCoroutineScope()
    val loader = remember { SharedImageLoader(KtorClientFactory().create()) }

    Box(
        modifier = modifier
            .height(108.dp)
            .fillMaxWidth()
            .clip(RoundedCornerShape(8.dp))
            .background(Color.White)
            .border(
                border = BorderStroke(1.dp, AppColor.Background.toComposeColor()), // Define border width and color
                shape = RoundedCornerShape(8.dp) // Define border shape (e.g., rounded corners)
            )
            .padding(0.dp)
    ) {
        // Bottom info bar
        Column(
            modifier = Modifier
                .align(Alignment.BottomStart)
                .fillMaxWidth(),
            verticalArrangement = Arrangement.Bottom,
            horizontalAlignment = Alignment.Start
        ) {
            Box(
                modifier = Modifier
                    .height(44.dp)
                    .fillMaxWidth()
                    .background(AppColor.Background.toComposeColor()),
            ) {
                Text(
                    text = name,
                    modifier = Modifier
                        .fillMaxWidth()
                        .wrapContentHeight(Alignment.CenterVertically)
                        .padding(start = 8.dp, top = 16.dp, bottom = 4.dp),
                    textAlign = TextAlign.Center,
                    style = FontUtils.body3.textStyle,
                    color = AppColor.GrayscaleDark.toComposeColor()
                )
            }
        }

        // Pokemon image
        if (bitmap != null) {
            Image(
                bitmap = bitmap!!.asImageBitmap(),
                contentDescription = name,
                modifier = Modifier
                    .size(60.dp)
                    .align(Alignment.Center),
                contentScale = ContentScale.Fit
            )
        } else {
            CircularProgressIndicator(
                modifier = Modifier
                    .size(60.dp)
                    .align(Alignment.Center)
            )

            // Load image from URL via SharedImageLoader
            LaunchedEffect(imageURL) {
                imageURL?.let { url ->
                    scope.launch {
                        loader.load(url)?.let { byteArray ->
                            val bmp = BitmapFactory.decodeByteArray(byteArray, 0, byteArray.size)
                            bitmap = bmp
                        }
                    }
                }
            }
        }

        // Top-right tag
        Row(
            modifier = Modifier
                .align(Alignment.TopEnd)
                .padding(top = 6.dp, end = 10.dp),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(2.dp)
        ) {
            Icon(
                painter = painterResource(id = R.drawable.tag),
                contentDescription = "tag",
                tint = AppColor.GrayscaleMedium.toComposeColor(),
                modifier = Modifier.size(12.dp)
            )
            Text(
                text = idTag.padStart(3, '0'),
                fontSize = 12.sp,
                color = AppColor.GrayscaleMedium.toComposeColor()
            )
        }
    }
}

@Preview
@Composable
fun AppAndroidPreview() {
    PokemonCard(
        name = "Bulbasaur",
        imageURL = "",
        idTag = "0"
    )
}